### ISSUE TITLE
[SMTChecker] ICE in Z3Interface::toZ3Expr(): "Sorts Int and bytes_tuple are incompatible" when indexing a mapping with bytes4 constant

### ISSUE THREAD

**Issue by cameel** [ethereum/solidity#10986]
From https://github.com/ethereum/solidity/pull/10961#issuecomment-782000639

## Description
This contract triggers an ICE in `soltest` when placed under `tests/smtCheckerTests/`:

```solidity
pragma experimental SMTChecker;

contract MockContract {
	bytes4 public constant SENTINEL_ANY_MOCKS = hex"01";
	mapping(bytes4 => bytes4) methodIdMocks;

	constructor() {
		methodIdMocks[SENTINEL_ANY_MOCKS] = 0;
	}
}
```

Output:
```
/solidity/test/boostTest.cpp(123): error: in "syntaxTests/multiSource/mock-contract": Exception during extracted test: /solidity/libsmtutil/Z3Interface.cpp(270): Throw in function z3::expr solidity::smtutil::Z3Interface::toZ3Expr(const solidity::smtutil::Expression&)
Dynamic exception type: boost::wrapexcept<solidity::smtutil::SMTLogicError>
std::exception::what: Sorts Int and bytes_tuple are incompatible
[solidity::util::tag_comment*] = Sorts Int and bytes_tuple are incompatible
```

The code comes originally from `mock-contract/contracts/MockContract.sol` in `gnosis-safe`. Extracted and slightly modified to remove all elements that are not relevant to the issue.

## Environment

- Compiler version: `0.8.2` (from `develop`).

**Comment by bshastry** [comment]
A few more examples synthesized by the fuzzer

```
pragma experimental SMTChecker;
contract e{
  function f(uint[] calldata) internal {}
  function h(uint[] calldata c) external { f(c[:]); }
}
```

throws

```
SMT logic error during analysis:
/solidity/libsmtutil/Z3Interface.cpp(270): Throw in function z3::expr solidity::smtutil::Z3Interface::toZ3Expr(const solidity::smtutil::Expression &)
Dynamic exception type: boost::wrapexcept<solidity::smtutil::SMTLogicError>
std::exception::what: Sort mismatch at argument #7 for function (declare-fun summary_10_function_f__8_21_0
             (Int
              Int
              abi_type
              crypto_type
              tx_type
              state_type
              uint[]_tuple
              state_type
              uint[]_tuple)
             Bool) supplied sort is bytes_tuple
[solidity::util::tag_comment*] = Sort mismatch at argument #7 for function (declare-fun summary_10_function_f__8_21_0
             (Int
              Int
              abi_type
              crypto_type
              tx_type
              state_type
              uint[]_tuple
              state_type
              uint[]_tuple)
             Bool) supplied sort is bytes_tuple
```

**Comment by bshastry** [comment]
Several more can be found by forcing SMT checker (by adding a pragma) to existing semantic/syntax tests. Here's one more

```
contract C {
int[] s;
function f(int[] calldata b, uint256 start, uint256 end) public returns (int) {
    s = b[start:end];
    uint len = end - start;
    assert(len == s.length);
    for (uint i = 0; i < len; i++) {
        assert(b[start:end][i] == s[i]);
    }
    return s[0];
}
}
// ====
// compileViaYul: also
// ----
// f(int256[], uint256, uint256): 0x60, 1, 3, 4, 1, 2, 3, 4 -> 2
```

throws

```
terminate called after throwing an instance of 'boost::wrapexcept<solidity::smtutil::SMTLogicError>'
  what():  Sorts uint[]_tuple and bytes_tuple are incompatible
```

**Comment by bshastry** [comment]
One more

```
pragma abicoder v1;
struct Item {
uint x;
uint y;
}

contract D {
Item[][][] public items;

function test() public view returns (uint) {
    // The autogenerated getters to not use ABI encoder.
    (uint a, uint b) = this.items(1, 2, 3);
    return a + b;
}
}
// ----
```

throws

```
terminate called after throwing an instance of 'boost::wrapexcept<solidity::smtutil::SMTLogicError>'
  what():  Sorts tuple(uint256,uint256) and struct Item are incompatible
```

**Comment by leonardoalt** [comment]
@bshastry the issues you reported seem unrelated to the original one? Could you please open new issues for them? (if that's not already the case and I missed them)

**Comment by bshastry** [comment]
> @bshastry the issues you reported seem unrelated to the original one? Could you please open new issues for them? (if that's not already the case and I missed them)

@leonardoalt I did not open a separate issue since they looked like duplicates at first glance. I created separate issues for each of the above examples (see https://github.com/ethereum/solidity/issues/11063 https://github.com/ethereum/solidity/issues/11064 https://github.com/ethereum/solidity/issues/11065)

**Comment by blishko** [comment]
I think this issue can be closed now, after the other examples have been moved to separate issues.
The original problem here has been fixed in #11052.
